### PR TITLE
Feat/deliverable4

### PR DESCRIPTION
**Pull request by FranciscoJRodriguezE**
## Description

- Refactor components to use Global State with Redux instead of Context/useReducer.
- Implement Add to Cart functionality
- Implement Shopping Cart Icon with badger indicating the number of items in the cart
- Implement Shopping Cart Page showing the items currently in the cart, subtotals and total.

- Link: https://franciscojrodrigueze.github.io/wa-react-2021-q4-internal-capstone-project/

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Refactor (non-breaking refactor which doesn't change functionality)
- [x] Styles (Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc))

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [x] I have added test instructions that prove my fix is effective or that my feature works
- [x] I have provided screenshots or recordings of results
- [x] I have added unit tests for the piece of work

# Evaluation Criteria
- [x] After clicking on the “Add to Cart” button on the Product Detail Page, the number of items selected are correctly added to the cart and the badge for the Shopping Cart Icon in the Header is updated correctly (25 points)
- [x] The logic to update the quantity for a product on the “Shopping Cart Page” is implemented correctly (10 points)
- [x] The logic to remove a product from the cart on the “Shopping Cart Page” is implemented correctly (10 points)
- [x] The form on the “Checkout Page” is displayed correctly (10 points)
- [x] The Order Summary table on the “Checkout Page” is displayed correctly (15 points)
- [x] The “Add to Cart” functionality is implemented correctly using React Context or Redux across all pages (30 points)

Bonus
- [x] PropTypes are implemented correctly for all your components (10 points)
- [ ] Error Boundaries are applied correctly at least for 1 component within the app (15 points)
- [ ] useMemo and useCallback hooks are implemented correctly to improve performance if needed (20 points)

# Notes 

This deliverable is incomplete. It lacks the Checkout Page.

### CODE REVIEW

File: src/components/ShoppingCartIcon/ShoppingCartIcon.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCartIcon/ShoppingCartIcon.component.jsx:7]: It can be a great idea to change this component to a generic icon or icon button component where the icon is received as a prop. Maybe not that useful in this project but just to practice is a great idea

File: src/components/ShoppingCartIcon/ShoppingCartIcon.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCartIcon/ShoppingCartIcon.component.jsx:21]: Also remember to always try to separate styles from the component itself to make the component more easily readable and maintainable

File: src/components/Grid/Grid.js
Comment (danbenitezf) [inline src/components/Grid/Grid.js]: To avoid confusion its better to rename the prop className, it may be something similar but i prefer to not name objects similar or equal  as 'reserved' words

File: src/components/Grid/Grid.js
Comment (danbenitezf) [inline src/components/Grid/Grid.js]: maybe handle this evaluation on a variable. here you can have just a variable that has the value of `pagination && totalPages > 1 `
```suggestion
                showPagination &&
```

File: src/components/GridCard/GridCard.js
Comment (danbenitezf) [inline src/components/GridCard/GridCard.js]: Handling this onClick event on a function on and just calling it here is better imo. 
```suggestion
                onClick={ handleOnClick }></input>
```
and the fn something like this
```
const handleOnClick = () => { // there is no need of adding properties because you are not passing any directly
   // then inside here 
   addToCart({
      id: cardId, 
      quantity: parseInt(1), // *
      price, // as the property has the same name as value you can leave it like this, also if you rename cardId to id 
                  you can do this. 
      stock: parseInt(stock)} // **
)};
```
`*`  I have to ask, why passing the number through parsInt? as you already have a number directly in the property the parseInt is not achieving anything
`* *` does the property comes from a string?

File: src/components/GridPaginator/GridPaginator.component.jsx
Comment (danbenitezf) [inline src/components/GridPaginator/GridPaginator.component.jsx]: Something extra you can handle for the paginator is disabling o not showing prev, next, first, last page depending on the current page. is you are on the first or last don't show the prev or next button. Totally fine if you don't do it because there is not enough time

File: src/components/PageRouter/PageRouter.component.jsx
Comment (danbenitezf) [inline src/components/PageRouter/PageRouter.component.jsx]: imo it is better to leave the definition of the routes alone. And in the parent component where you are using them handle what you have below to switch from home to products.
Something like this
image
A router component as i mentioned is typically used just for the definition of the routes
```suggestion
const PageRouter = () => {
    
  return (
    <>
      <Switch>
        <Route exact path="/home" component={Home} />
        <Route exact path="/products" component={ProductList} />
        <Route exact path="/product/:productId" component={ProductDetail} />
        <Route exact path='/search' component={SearchResults} />
        <Route exact path='/cart' component={ShoppingCart} />
        <Route path="*">
          <span>404 Not Found.</span>
        </Route>
      </Switch>
    </>
  );
}

export default PageRouter;
```

Comment (danbenitezf) [reply src/components/PageRouter/PageRouter.component.jsx]: You can even use a map and then on a separate file define the routes. IE:
create a routes config file `routes-config.js`: 
```
export const routes = [
  { path: '/', component: Home, exact: true },
  { path: '/login', component: Login, exact: true },
  { path: '*', component: Home, exact: true },
];
```
and on the router itself
```
export const PageRouter = () => {
    
  return (
    <>
      <Switch>
       {
          routes.map({ route, component }) =>  {
            <Route exact path={route} component={component} />
         })
       }
        <Route path="*">
          <span>404 Not Found.</span>
        </Route>
      </Switch>
    </>
  );
};
```

File: src/components/ProductDetail/ProductDetail.component.jsx
Comment (danbenitezf) [inline src/components/ProductDetail/ProductDetail.component.jsx:26]: I think the evaluation its okay here maybe store it in a variable and then the return of the loading on the return of the component

File: src/components/SearchResults/List/List.component.jsx
Comment (danbenitezf) [inline src/components/SearchResults/List/List.component.jsx]: Maybe this can be modified to something like this. Sorry in advance if there is an error in the syntax.
```suggestion
     return (
     {
        parsedData.length > 0 ? 
        <div className={className}>
            <ListPaginator /> 
                // now that the paginator works correctly i think its okay if you leave it even if there is only 
                // one page but it is up to you
            <div className="list-cards">
                {
                    parsedData.map((data) => {
                        return (<ListCard key={data.id} data={data} />)
                    })
                }
            </div>
            <ListPaginator />
        </div>
        : 
        <h4 style={{textAlign: "center"}}>{noElementsCustomMessage}</h4>
     }
        
    )
```

File: src/components/SearchResults/List/ListCard/ListCard.component.jsx
Comment (danbenitezf) [inline src/components/SearchResults/List/ListCard/ListCard.component.jsx]: For me it really is better to have what is done `onSomething` functions on its own function, it is easy to maintain

File: src/components/SearchResults/List/ListPaginator/ListPaginator.component.jsx
Comment (danbenitezf) [inline src/components/SearchResults/List/ListPaginator/ListPaginator.component.jsx]: may be a good idea to move the function to a utils function or a hook and then only use it here

File: src/components/SearchResults/List/ListPaginator/ListPaginator.component.jsx
Comment (danbenitezf) [inline src/components/SearchResults/List/ListPaginator/ListPaginator.component.jsx]: just remember to remove this commented block of code

File: src/components/ShoppingCart/ShoppingCart.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCart/ShoppingCart.component.jsx]: i think its better to move this to a fn and then show here only the total price

File: src/components/ShoppingCart/ShoppingCartTable/ShoppingCartCard/ShoppingCartCard.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCart/ShoppingCartTable/ShoppingCartCard/ShoppingCartCard.component.jsx]: Is `quantity2` used?

File: src/components/ShoppingCart/ShoppingCartTable/ShoppingCartCard/ShoppingCartCard.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCart/ShoppingCartTable/ShoppingCartCard/ShoppingCartCard.component.jsx]: I think this should be moved on to its own function and there you could also update the subtotal
```suggestion
                <select defaultValue={quantity} onInput={ handleQuantityonInput }>
```

File: src/components/ShoppingCart/ShoppingCartTable/ShoppingCartTable.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCart/ShoppingCartTable/ShoppingCartTable.component.jsx]:
```suggestion
        return itemFound ? { ...element, quantity: itemFound.quantity } : { ...element};
        // when you do this the conditional or ternary operator evaluates that that object you are passing is not null, 
        // undefined, false or 0 (as in js 0 is false), it helps a lot to reduce code
```

File: src/hooks/useProductById.js
Comment (danbenitezf) [inline src/hooks/useProductById.js]: If you move this on to a  utils file that only stores the fn would be awesome to have the file as clear as possible

File: src/state/reducers/ShoppingCartReducer.js
Comment (danbenitezf) [inline src/state/reducers/ShoppingCartReducer.js:16]: it may be clearer if you move this little section to a variable and then only assign it here

**Review by danbenitezf — APPROVED** [review]
Good job. If there is anything or a specific file or section of the application you would like to have a second review just ask me. 
You have only two criterias left so try to work on them

File: src/components/GridCard/GridCard.js
Comment (FranciscoJRodriguezE) [reply src/components/GridCard/GridCard.js]: The property was being considered a string inside the reducer for some reason and was throwing NaN when computed by the addToCart action. I will consider moving the conversion inside the reducer.

Comment (FranciscoJRodriguezE) [reply src/components/GridCard/GridCard.js]: I've found the issue with the NaN: I was not getting the "stock" property from the API call. Oops. Removing parseInt's

File: src/components/ProductDetail/ProductDetail.component.jsx
Comment (FranciscoJRodriguezE) [reply src/components/ProductDetail/ProductDetail.component.jsx:26]: That if and return is guarding against the empty array useProductById returns while the API call is loading (isLoading === true). If I apply the suggested refactoring, then it's gonna crash.

**Comment by FranciscoJRodriguezE** [comment]
All comments have been resolved. Working on the Checkout page from here.

**Comment by FranciscoJRodriguezE** [comment]
This deliverable is complete. I would like a final review before next week.

File: src/App.js
Comment (danbenitezf) [inline src/App.js]: I don't remember if i've suggested you this, sorry if I haven't mentioned it before. 
IMO it is a lot better to split this component into 3 child components `Header`, `Content` and `Footer` not styled components, but functional components. 
```
const App = () => {
    return (
      <Container>
        <Header/>
        <Content/>
        <Footer/>
      </Container>
    );
}
```
You already have most already, its just dividing it into the corresponding components

Comment (danbenitezf) [reply src/App.js]: This way is easy to make changes to just one section

File: src/components/Checkout/Checkout.component.jsx
Comment (danbenitezf) [inline src/components/Checkout/Checkout.component.jsx:80]: My suggestion here is to change it to a little bit different structured component. Use the `Checkout.component.jsx` just as a container for the real components that handle something very specific for the page. Try to think this way for everything in the app.
```suggestion
  return (
    // i know classes may need to be changed a little bit to reflect the changes.
    <div className="checkout-container">
       // here goes the form, not all of what you have on the component needs to be part of the form
       <CheckoutForm /> 
       // Then a component for your summary for the checkout
       <CheckoutSummary />
       // Then you can have your action buttons for placing the order or go back 
       <CheckoutAction />
    <div/>
  );
};
```

Comment (danbenitezf) [reply src/components/Checkout/Checkout.component.jsx:80]: For the checkout form I want to add a deeper suggestion. For forms that are just very repetitive and that you know that have the same structure i'll do something like this. 
Create a simple input component `FormInput.jsx`
```
export const FormInput = ( {name, type, label }) => {
// inside this component then you can even handle the input change and more but for what is required for the 
// deliverable this is fine
   return (
      <div className="input>
         <label htmlFor={name}> {label} </label>
          <input name={name} type={type}></input>
      <div />
   );
}
```

then your checkout form would be

```
export const CheckoutForm = () => {
   // An array that consists of all the necessary inputs
   const checkoutInputs = [ { name:'name', type: 'text', label: 'Customer Name' } ];
   return (
      <>
          {
            checkoutInputs.map( (input) =>  (
                <FormInput {...input} />
            ))
          }
      </>
   );
}
```
For this project there is not a very big form or a lot that you have but believe me when there are projects with forms that have 20, 30 or more inputs and are all the same having this kind of structure helps a lot to maintain and make changes

File: src/components/GridPaginator/GridPaginator.component.jsx
Comment (danbenitezf) [inline src/components/GridPaginator/GridPaginator.component.jsx]: As you are not returning anything in the 'else'
```suggestion
            {   
                (currentPage !== 1) && 
                <>
                    <button onClick={() => first() }>First</button> 
                    <button onClick={() => prev() }>Prev</button>
                 </> 
```

File: src/components/GridPaginator/GridPaginator.component.jsx
Comment (danbenitezf) [inline src/components/GridPaginator/GridPaginator.component.jsx]: Same here for the 'else'
```suggestion
            {   
                (currentPage !== totalPages) && 
                <>
                    <button onClick={() => next() }>Next</button>
                    <button onClick={() => last() }>Last</button> 
                  </>
```

File: src/components/PageRouter/PageRouter.component.jsx
Comment (danbenitezf) [inline src/components/PageRouter/PageRouter.component.jsx]: I think a better name would be `AppRouter`

File: src/components/PageRouter/PageRouter.component.jsx
Comment (danbenitezf) [inline src/components/PageRouter/PageRouter.component.jsx]: If you have this on a separate file in case the route grow to much so its easy to maintain (I don't believe its gonna happen on this project but not a bad idea to do it)

File: src/components/ProductDetail/ProductDetail.component.jsx
Comment (danbenitezf) [inline src/components/ProductDetail/ProductDetail.component.jsx:57]: Maybe create a component for the label and do a map of the labels you need or just map the labels from an array of labels

File: src/components/SearchResults/List/List.component.jsx
Comment (danbenitezf) [inline src/components/SearchResults/List/List.component.jsx:31]: I think is better to create a class for this style and avoid having the style directly on the element

File: src/components/ShoppingCart/ShoppingCart.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCart/ShoppingCart.component.jsx:44]: I believe this two divs are not doing much
```suggestion
            <ShoppingCartTable />
```

File: src/components/ShoppingCart/ShoppingCart.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCart/ShoppingCart.component.jsx]: and it is the same for this one
```suggestion
    const proceedToCheckout = () => {
```

File: src/components/ShoppingCart/ShoppingCart.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCart/ShoppingCart.component.jsx]: i don't think you need to pass `items` to the function as you have items at the root of your component it is accessible through all your functions
```suggestion
    const calculateTotalPrice = () => {
```

File: src/components/ShoppingCart/ShoppingCart.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCart/ShoppingCart.component.jsx]:
```suggestion
            <div className="shoppingCart-total">Total: ${ calculateTotalPrice }</div>
            <div className="shoppingCart-checkout"><button onClick={proceedToCheckout }>Checkout</button></div>              
```

File: src/components/ShoppingCart/ShoppingCart.component.jsx
Comment (danbenitezf) [inline src/components/ShoppingCart/ShoppingCart.component.jsx]: For this i think i would have used an useEffect 
```suggestion
    const { totalPrice, setTotalPrice } = useState(0);
    useEffect(() => {
       setTotalPrice(items.reduce( (previous, current) => previous + (current.price*current.quantity),0));
    }, [items]);
```
on line `45` 
```
<div className="shoppingCart-total">Total: ${ totalPrice }</div>
```
and on `33`
```
const checkoutData = { orderSummary, total: totalPrice}
```

File: src/components/GridPaginator/GridPaginator.component.jsx
Comment (danbenitezf) [inline src/components/GridPaginator/GridPaginator.component.jsx:12]: Good job 👍

**Review by danbenitezf — COMMENTED** [review]
Good Job. Any help needed for next deliverable you can ping me on slack if you want

**Review by danbenitezf — APPROVED** [review]
LGTM go ahead with the next deliverable

**Pull request merged**
